### PR TITLE
[filesys] Add reporting of ext fragmentation with e2freefrag

### PR DESCRIPTION
**Pull request by Saruspete**
Hello,

I've added an option for e2freefrag reporting from filesys plugin. The binary should be part of e2fsprogs package in most distributions.

### CODE REVIEW

File: sos/plugins/filesys.py [hunk 1]
```diff
@@ -53,6 +54,10 @@ def setup(self):
         for dev in self.do_regex_find_all(ext_fs_regex, mounts):
                 self.add_cmd_output("dumpe2fs %s %s" % (dumpe2fs_opts, dev))
 
+        if self.get_option('frag'):
+            for dev in self.do_regex_find_all(ext_fs_regex, mounts):
+                self.add_cmd_output("e2freefrag %s" % (dev))
```
Comment (bmr-cymru) [inline sos/plugins/filesys.py]: Why do this loop again here? Just add a conditional on `self.get_option('frag')` inside the existing loop over `ext2` family file systems.

File: sos/plugins/filesys.py [hunk 2]
```diff
@@ -25,7 +25,7 @@ class Filesys(Plugin, RedHatPlugin, DebianPlugin, UbuntuPlugin):
     option_list = [("lsof", 'gathers information on all open files', 'slow',
                     False),
                    ("dumpe2fs", 'dump filesystem information', 'slow', False),
-                   ("frag", 'filesystem fragmentation status', 'slow', False) ]
+                   ("frag", 'filesystem fragmentation status', 'slow', False)]
```
Comment (bmr-cymru) [inline sos/plugins/filesys.py]: This block wasn't formatted properly when you started, but we should make things better, not worse :-)

The accepted style for option lists in `sos` (due to their length) is to have the closing `]` on a line of its own, and (optionally) to place a newline between the opening `[` and the first option opening `(`, i.e.:

```python
    option_list = [
        ("opt1", "description for opt1", "fast", False),
        ("opt2, "description for opt2", "fast", False)
    ]
```

This gives the most space available to the option line, conforms to `pep8`, and makes the list more readable.

**Comment by bmr-cymru** [comment]
The changes seem mostly fine (modulo the option list formatting and commit messages), although I suspect they won't get much use as they require the user to specify an option.

Have you done any measurements to determine the actual cost of running with the extra commands?

You also need to re-work your commits to comply with the [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines): you're missing a Signed-off-by (DCO), and you need to squash the fixup commit into the first - either use `git rebase -i`, or `git reset HEAD~1` and `git commit --amend`. You can then re-push the branch in the PR with `git push --force`.

**Comment by Saruspete** [comment]
Good points, thanks for the procedure :)
Is it ok now ?

Regarding the time, it took 1 second on a 510G-Free / 670G raid5, and 5 seconds on a 23G-Free/207G RAID1.
It seems to depend on the used space and underlying devices.

<details>
<summary>23G-Free/207G RAID1 </summary>
sudo /usr/bin/time -v e2freefrag /dev/mapper/rootvg-data

Device: /dev/mapper/rootvg-data
Blocksize: 4096 bytes
Total blocks: 55050240
Free blocks: 8545508 (15.5%)

Min. free extent: 4 KB 
Max. free extent: 63220 KB
Avg. free extent: 4124 KB

HISTOGRAM OF FREE EXTENT SIZES:
Extent Size Range :  Free extents   Free Blocks  Percent
    4K...    8K-  :          1200          1200    0.01%
    8K...   16K-  :           730          1736    0.02%
   16K...   32K-  :          1498          7649    0.09%
   32K...   64K-  :           291          3167    0.04%
   64K...  128K-  :           247          5534    0.06%
  128K...  256K-  :           229         10626    0.12%
  256K...  512K-  :           247         22755    0.27%
  512K... 1024K-  :           327         59879    0.70%
    1M...    2M-  :           553        213756    2.50%
    2M...    4M-  :           715        531236    6.22%
    4M...    8M-  :          1255       1840684   21.54%
    8M...   16M-  :           335        888486   10.40%
   16M...   32M-  :           403       2304139   26.96%
   32M...   64M-  :           256       2654661   31.06%
	Command being timed: "e2freefrag /dev/mapper/rootvg-data"
	User time (seconds): 0.34
	System time (seconds): 0.01
	Percent of CPU this job got: 6%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:05.92
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 7548
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 1954
	Voluntary context switches: 1676
	Involuntary context switches: 361
	Swaps: 0
	File system inputs: 13592
	File system outputs: 0
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
</details>

<details>
<summary>510G-Free / 670G raid5</summary>
sudo /usr/bin/time -v e2freefrag /dev/mapper/data-data

Device: /dev/mapper/data-data
Blocksize: 4096 bytes
Total blocks: 178420736
Free blocks: 129931522 (72.8%)

Min. free extent: 4 KB 
Max. free extent: 2064256 KB
Avg. free extent: 903196 KB

HISTOGRAM OF FREE EXTENT SIZES:
Extent Size Range :  Free extents   Free Blocks  Percent
    4K...    8K-  :            18            18    0.00%
    8K...   16K-  :             4             9    0.00%
   16K...   32K-  :             1             6    0.00%
   32K...   64K-  :             5            56    0.00%
   64K...  128K-  :             4            79    0.00%
  128K...  256K-  :             5           210    0.00%
  256K...  512K-  :            11          1053    0.00%
  512K... 1024K-  :            13          2323    0.00%
    1M...    2M-  :            14          5032    0.00%
    2M...    4M-  :            42         34818    0.03%
    4M...    8M-  :            32         47167    0.04%
    8M...   16M-  :             1          2097    0.00%
   16M...   32M-  :             5         29744    0.02%
   32M...   64M-  :             7         64511    0.05%
   64M...  128M-  :           187       4488608    3.45%
  256M...  512M-  :             2        210879    0.16%
  512M... 1024M-  :             9       1782680    1.37%
    1G...    2G-  :           275     136713313  105.22%
	Command being timed: "e2freefrag /dev/mapper/data-data"
	User time (seconds): 1.04
	System time (seconds): 0.01
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:01.05
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 22604
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 5750
	Voluntary context switches: 1
	Involuntary context switches: 1062
	Swaps: 0
	File system inputs: 0
	File system outputs: 0
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
</details>

**Comment by bmr-cymru** [comment]
OK, thanks for confirming (and apologies for the slow response). I'll keep this behind the option for now unless we get a request to make it default.

**Pull request closed without merging**